kube/terraform/main.tf

```diff
@@ -0,0 +1,309 @@
+  }
+}
+
+resource "kubernetes_service" "etcd_service" {
+  metadata {
+    name      = "etcd"
+    namespace = "m3db"
+    labels {
+      app = "etcd"
+    }
+  }
+  spec {
+    port {
+      name = "client"
+      port = 2379
+    }
+    port {
+      name = "peer"
+      port = 2380
+    }
+    selector {
+      app = "etcd"
+    }
+    cluster_ip = "None"
+  }
+}
+
+resource "kubernetes_service" "etcd_cluster" {
+  metadata {
+    name      = "etcd-cluster"
+    namespace = "m3db"
+    labels {
+      app = "etcd"
+    }
+  }
+  spec {
+    port {
+      protocol = "TCP"
+      port     = 2379
+    }
+    selector {
+      app = "etcd"
+    }
+    type = "ClusterIP"
+  }
+}
+
+resource "kubernetes_stateful_set" "etcd_stateful_set" {
+  metadata {
+    name      = "etcd"
+    namespace = "m3db"
+    labels {
+      app = "etcd"
+    }
+  }
+  spec {
+    service_name = "etcd"
+    replicas = 3
+    selector {
+      match_labels {
+        app = "etcd"
+      }
+    }
+    template {
+      metadata {
+        labels {
+          app = "etcd"
+        }
+      }
+      spec {
+        container {
+          name    = "etcd"
+          image   = "quay.io/coreos/etcd:v3.3.3"
+          command = ["etcd", "--name", "$(MY_POD_NAME)", "--listen-peer-urls", "http://$(MY_IP):2380", "--listen-client-urls", "http://$(MY_IP):2379,http://127.0.0.1:2379", "--advertise-client-urls", "http://$(MY_POD_NAME).etcd:2379", "--initial-cluster-token", "etcd-cluster-1", "--initial-advertise-peer-urls", "http://$(MY_POD_NAME).etcd:2380", "--initial-cluster", "etcd-0=http://etcd-0.etcd:2380,etcd-1=http://etcd-1.etcd:2380,etcd-2=http://etcd-2.etcd:2380", "--initial-cluster-state", "new", "--data-dir", "/var/lib/etcd"]
+          port {
+            name           = "client"
+            container_port = 2379
+          }
+          port {
+            name           = "peer"
+            container_port = 2380
+          }
+          env {
+            name = "MY_IP"
+            value_from {
+              field_ref {
+                field_path = "status.podIP"
+              }
+            }
+          }
+          env {
+            name = "MY_POD_NAME"
+            value_from {
+              field_ref {
+                field_path = "metadata.name"
+              }
+            }
+          }
+          env {
+            name  = "ETCDCTL_API"
+            value = "3"
+          }
+          volume_mount {
+            name       = "etcd-data"
+            mount_path = "/var/lib/etcd"
+          }
+        }
+      }
+    }
+    volume_claim_template {
+      metadata {
+        name = "etcd-data"
+      }
+      spec {
+        access_modes = ["ReadWriteOnce"]
+        resources {
+          requests {
+            storage = "50Gi"
+          }
+        }
+        storage_class_name = "fast"
+      }
+    }
+  }
+}
+
+resource "kubernetes_config_map" "m3dbnode_config" {
+  metadata {
+    name      = "m3dbnode-config"
+    namespace = "m3db"
+  }
+  data {
```

Comment:
Would it make sense to reference a [template file](https://www.terraform.io/docs/providers/template/d/file.html) that uses the [configuration map](https://github.com/m3db/m3/blob/master/kube/m3dbnode-configmap.yaml) already located within the `kube` directory?

Reply:
Hmm, it would make this sub package depend on the top level file. I think if we were to publish this in the terraform public repo, everything has to be self contained.

Reply:
It does mean this may need to be in sync with the latest config, but I would hope we don't make backward breaking changes that often.

Reply:
Yea, that ugly string seems like a PITA to update if it does break it. It's kinda of a PITA already to ensure that _all_ yaml configs are updated let alone a yaml => string like this. Will leave the decision in your hands.